### ISSUE TITLE
Question: Using NNG with libevent

### ISSUE THREAD

**Issue by ravinathwani** [nanomsg/nng#840]
We have a distributed app that needs a low-latency messaging framework. We have chosen going with NNG and I have gone through a lot of documents and man pages to understand how to integrate it with our app.
Our app tasks are single-threaded and uses libevent for other sockets, timers, signal-handlers etc. and I see some restrictions using NNG with libevent (gone through some issues/mailing lists related to this).
We plan to integrate the Req/Rep protocol where
- Task A sends a request to task B and expects an async reply
- There could be multiple outstanding requests on task A
- When B replies, A needs to perform req-to-reply matching
I went though the NNG 'context' docs to understand it and seems it can be used here, but then it clearly mentions that use of file-descriptor polling while contexts are in use on the same socket is not supported.

My understanding is as follows
The current app is single threaded and libevent runs in main thread serving other events.
Integrate Contexts and AIO to support async processing of multiple requests (which will result in multiple threads) and DO NOT add the NNG socket fd to libevent.
Questions in this approach:
1. How to make sure NNG framework does not impact any of the other libevent handling?
2. Would the msg receive path do processing in the thread in which it was received? (may result in either sending a reply or sending a request to another task waiting for its response)
3. In the main thread, while other socket read/write or timer events occur, it may result in performing a msg send, would this cause any issue?
4. In general, is there a way to control which threads run the NNG msg framework?
Also, I believe we would need some sync/locking if same data structures are accessed via main thread running libevent and other threads handling NNG req/rep messages.

Is there any other way where NNG sockets and contexts/aio integration can be made simpler with libevent?

Thanks & Regards,
Ravi Nathwani

**Comment by gdamore** [comment]
NNG uses it's own event loops (depending on the platform) for processing I/O, and then calls your completion handler on a separate thread that is created for processing completions.  (Although, in some circumstances we complete the operation without a context switch doing it on the I/O thread -- but that is an optimization that is only done when we can be certain that it is safe to do so.)

You generally should not worry about the threads that NNG uses.  (Having said that, if you have extreme low latency requirements -- high frequency trading style latencies of 10usec for example -- then this does matter, and supporting that kind of application should probably be done in collaboration with detailed NNG expertise -- send me a message if this applies to you.)

The one significant rule you need to follow is that completion callback threads must not block, or attempt to acquire locks that might be held by other threads that block.  If you follow those rules, then everything else should Just Work.

Note that specifically it is *legal* and common to use non-blocking message send or receive (such as other functions using AIOs) from the completion callbacks.

I suspect that this will work well with libevent.

Having said all that, I would be curious to know if you save much by avoiding the file descriptors for libevent.  The reason for this is that libevent may actually fabricate file descriptors and alert on them to wake it's own poller, in which case the cost should be about the same as what NNG would do if it fabricated the descriptors for you instead of libevent doing so.

**Comment by ravinathwani** [comment]
Thanks Garrett for your quick response !

Our app today is a functional prototype consisting of multiple single threaded Linux processes that use non-blocking UDP socket based IPC on the loopback IP to talk to each other. These processes perform different kinds of functionality including external network I/O, timers, signal handling, syslog etc. (no disk I/O) and do not have any blocking calls and all these functions are async/event-driven using libevent with plain C code.

This app would eventually run on a distributed virtualized env with processes running on different VMs/containers. For task/service discovery we plan to use DNS based discovery and need a distributed IPC messaging framework that is reliable, low-latency (approx 50-100usec), support name resolutions, provides connection management and msg auto-retry etc. After evaluating a few broker-based and brokerless frameworks, finally we think NNG is most suitable to meet our needs and your response gives more confidence.

A few follow-up questions
1. Should the nng_req_open/nng_rep_open calls be done in the main thread or have a pthread_create and call these within the pthread?
2. There is no blocking code today in the app and ideally we want to keep it that way. But with different NNG threads, if multiple messages are being processed simultaneously in different threads, this processing might need some common app data structures that would introduce some kind of sync requirements between these threads. Though the app needs the ability to support multiple outstanding requests, there is no requirement to have simultaneous message processing in different threads. Would this be possible?
3. If we integrate the NNG socket fd with libevent, will it address the above issue (#2)?
   Does NNG polling ensure picking up a single message on one alert/wake-up from libevent OR
   Does NNG polling pick-up all messages available on the NNG socket fd and call their handlers in different processing threads on one alert/wake-up from libevent?
4. Although your reply seems to be positive, I wanted to understand the reason behind not supporting multiple contexts on same socket with libevent. Does nng_ctx_open result in a new thread within?
5. For supporting multiple outstanding requests on a req socket via contexts, the nng_ctx example creates a nng_aio_alloc per nng_ctx_open. Given our application understanding, do we need this? Sorry, but I am not able to visualize ctx/aio/multithreading and libevent together.
6. I could not find any architecture diagram/description of NNG internals. I think that would really help many people to understand how to correctly integrate NNG within their application.

Thanks again,
Ravi Nathwani

**Comment by ravinathwani** [comment]
Thanks Garrett for your detailed responses.

Regarding the second concern, let me explain via an example.
- Task A opens a req socket and opens multiple contexts on this socket
- Task B opens a rep socket (does it need to create as many contexts manually/automatically?)
- Task A sends multiple requests to Task B simultaneously (each in a different context) and waits for responses
- Task B main thread libevent loop wakes-up and calls a callback for NNG socket fd (which NNG API needs to be called here?)
- Is the NNG I/O thread the main thread where libevent runs?
- Task B NNG I/O thread reads each received request and calls a completion handler on a separate thread
- Does this mean multiple completion handlers run in parallel in different threads?
- If above is true, do we need to make our completion handlers (and hence most part of the app) thread-safe?
- Can the app completion handlers be always called in the main thread?

In general, is it possible to integrate NNG in our app and continue to be single-threaded (i.e. everything running in main thread)?

**Comment by gdamore** [comment]
Under the hood, NNG uses threads.  You cannot prevent that.  Nor can you arrange what thread your callback handlers run on.  We generally run them from one of the internal NNG threads.

If you're wanting single-threaded behavior like you're suggesting here, you can use the non-blocking socket operations, along with NNG_OPT_SENDFD or NNG_OPT_RECVFD to obtain notification file descriptors.  This won't be as efficient, but it will let you pretend that NNG is running entirely single threaded.  (Meaning you won't have to worry about threading in your application.)

If you want to use callback handlers normally instead, then yes, you have to accept that multiple callbacks can run in parallel (there is a thread pool).  HOWEVER, a given aio can only have one operation pending, and thus only one callback running, at a time.  That can help with synchronization.

**Comment by gdamore** [comment]
I'm closing this, since there seems to be no actionable things here.  (Well, doc improvements maybe....)
